## Analysing population income UCI Adult

Link to HW task (original): https://github.com/Yorko/mlcourse.ai/blob/master/jupyter_russian/assignments_demo/assignment01_adult_pandas.ipynb <br>
Information about attributes: <br>
**age**: continuous. <br>
**workclass**: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked. <br>
**fnlwgt**: continuous. <br>
**education**: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool. <br>
**education-num**: continuous. <br>
**marital-status**: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse. <br>
**occupation**: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces. <br>
**relationship**: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried. <br>
**race**: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black. <br>
**sex**: Female, Male. <br>
**capital-gain**: continuous. <br>
**capital-loss**: continuous. <br>
**hours-per-week**: continuous. <br>
**native-country**: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands. <br>
**salary**: >50K,<=50K <br>


In [45]:
import pandas as pd
import numpy as np

In [16]:
data=pd.read_csv("adult.data.csv")
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

*Let's see with what types of data do we operate. It will helps us in completing the tasks

In [27]:
data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
age               32561 non-null int64
workclass         32561 non-null object
fnlwgt            32561 non-null int64
education         32561 non-null object
education-num     32561 non-null int64
marital-status    32561 non-null object
occupation        32561 non-null object
relationship      32561 non-null object
race              32561 non-null object
sex               32561 non-null object
capital-gain      32561 non-null int64
capital-loss      32561 non-null int64
hours-per-week    32561 non-null int64
native-country    32561 non-null object
salary            32561 non-null object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


**1. Print out number of men and women in this dataset.**

In [25]:
data[data['sex'] == 'Female']['sex'].count(),  \
data[data['sex'] == 'Male']['sex'].count()

(10771, 21790)

In [26]:
data['sex'].value_counts() #another way to print it out

Male      21790
Female    10771
Name: sex, dtype: int64

**2. What is the average age of women?**

In [15]:
round(data[data['sex'] == 'Female']['age'].mean(),2)

36.86

**3. What is the proportion of Germans(data['native-country'] == 'Germany')?**
First of all, let's see how people are distributed across different countries

In [31]:
data['native-country'].value_counts() #let's first see the

United-States                 29170
Mexico                          643
?                               583
Philippines                     198
Germany                         137
Canada                          121
Puerto-Rico                     114
El-Salvador                     106
India                           100
Cuba                             95
England                          90
Jamaica                          81
South                            80
China                            75
Italy                            73
Dominican-Republic               70
Vietnam                          67
Guatemala                        64
Japan                            62
Poland                           60
Columbia                         59
Taiwan                           51
Haiti                            44
Iran                             43
Portugal                         37
Nicaragua                        34
Peru                             31
Greece                      

The maximum is in *United States* - 29170 while we have only one person from netherlands. 

In [28]:
data[data['native-country'] == 'Germany']['sex'].count()/data['sex'].count()*100

0.42074874850281013

**4-5. What are the mean values and standard deviations of the age of those who receive more than 50K per year (salary sign) and those who receive less than 50K per year?**

In [34]:
data['salary'].value_counts()

<=50K    24720
>50K      7841
Name: salary, dtype: int64

In [38]:
data[(data['salary'] == '<=50K')]['age'].mean(), \
data[(data['salary'] == '<=50K')]['age'].std()

(36.78373786407767, 14.020088490824813)

In [39]:
data[(data['salary'] == '>50K')]['age'].mean(), \
data[(data['salary'] == '>50K')]['age'].std()

(44.24984058155847, 10.51902771985177)

*From the numbers above it is seen that the mean age of people whose salary was below 50K(>=50K) is ~37 with standard deviation 14 years. While the mean age value of people with a salary higher than 50K (<50K) is 44 +- 10.5 years. The hypothesis from these observations could be that the wage of a person depends on the age. However, it definitely depends on other factors and figuring it out is not the point of this homework.*

**6. Is it true that people who receive more than 50k have at least a higher education? (education: Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters or Doctorate)**

In [105]:
data[data['salary'] == '>50K']['education'].unique()

array(['HS-grad', 'Masters', 'Bachelors', 'Some-college', 'Assoc-voc',
       'Doctorate', 'Prof-school', 'Assoc-acdm', '7th-8th', '12th',
       '10th', '11th', '9th', '5th-6th', '1st-4th'], dtype=object)

*No, it is not true. The least level of education from the people whose salary is higher than 50K is 1st-4th grade of the school.*

**7. Print out the age statistics for each race and each sex. What is the maximum age of Female that belongs to Amer-Indian-Eskimo?**

In [106]:
data.groupby(['race','sex'])['age'].describe()

count       mean        std   min   25%   50%  \
race               sex                                                       
Amer-Indian-Eskimo Female    119.0  37.117647  13.114991  17.0  27.0  36.0   
                   Male      192.0  37.208333  12.049563  17.0  28.0  35.0   
Asian-Pac-Islander Female    346.0  35.089595  12.300845  17.0  25.0  33.0   
                   Male      693.0  39.073593  12.883944  18.0  29.0  37.0   
Black              Female   1555.0  37.854019  12.637197  17.0  28.0  37.0   
                   Male     1569.0  37.682600  12.882612  17.0  27.0  36.0   
Other              Female    109.0  31.678899  11.631599  17.0  23.0  29.0   
                   Male      162.0  34.654321  11.355531  17.0  26.0  32.0   
White              Female   8642.0  36.811618  14.329093  17.0  25.0  35.0   
                   Male    19174.0  39.652498  13.436029  17.0  29.0  38.0   

                             75%   max  
race               sex                  
Amer-Indian-Eskimo Female  46.00  80.0  
                   Male    45.00  82.0  
Asian-Pac-Islander Female  43.75  75.0  
                   Male    46.00  90.0  
Black              Female  46.00  90.0  
                   Male    46.00  90.0  
Other              Female  39.00  74.0  
                   Male    42.00  77.0  
White              Female  46.00  90.0  
                   Male    49.00  90.0

*The maximum age of Female Amer-Indian-Eskimo is 80 years old.*

**8. Among which category the percentage of earning a lot (> 50K) is higher: among married or single men (a sign of marital-status)? We consider married people who have marital-status starts with Married (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), and consider the rest as single.** 

In [107]:
data[(data['sex'] == 'Male') &
     (data['marital-status'].isin(['Never-married', 
                                   'Separated', 'Divorced']))]['salary'].value_counts()

<=50K    7423
>50K      658
Name: salary, dtype: int64

In [108]:
data[(data['sex'] == 'Male') &
     (data['marital-status'].str.startswith('Married'))]\
    ['salary'].value_counts()

<=50K    7576
>50K     5965
Name: salary, dtype: int64

In [109]:
data['marital-status'].value_counts()

Married-civ-spouse       14976
Never-married            10683
Divorced                  4443
Separated                 1025
Widowed                    993
Married-spouse-absent      418
Married-AF-spouse           23
Name: marital-status, dtype: int64

*The salary is higher amount married male ;) DO NOT WAIT - DO A PROPOSAL!*

**9. What is the maximum number of hours a person works per week (sign of hours-per-week)? How many people work so many hours and what percentage of them earn more than 50K (>50K)?**

In [114]:
max_load = data['hours-per-week'].max()
print("Highest workload - {0} h./week".format(max_load))

num_workaholics = data[data['hours-per-week'] == max_load].shape[0]
print("Number of such hard working employees = {0}".format(num_workaholics))

rich_share = float(data[(data['hours-per-week'] == max_load)
                 & (data['salary'] == '>50K')].shape[0]) / num_workaholics
print("with salary higher than >50K - {0}%".format(int(100 * rich_share)))

Highest workload - 99 h./week
Number of such hard working employees = 85
with salary higher than >50K - 29%


**10. Calculate the average work time load based on salary for each country.**

In [116]:
data[(data['native-country'] == 'Japan')].groupby(['native-country','salary'])['hours-per-week'].describe()

count       mean        std   min   25%   50%   75%  \
native-country salary                                                        
Japan          <=50K    38.0  41.000000  11.902759  10.0  40.0  40.0  41.5   
               >50K     24.0  47.958333  16.120414  21.0  40.0  42.5  50.0   

                        max  
native-country salary        
Japan          <=50K   65.0  
               >50K    99.0

In [104]:
#another 
pd.crosstab(data['native-country'], data['salary'], 
           values=data['hours-per-week'], aggfunc=np.mean).T

native-country          ?   Cambodia     Canada      China   Columbia  \
salary                                                                  
<=50K           40.164760  41.416667  37.914634  37.381818  38.684211   
>50K            45.547945  40.000000  45.641026  38.900000  50.000000   

native-country       Cuba  Dominican-Republic    Ecuador  El-Salvador  \
salary                                                                  
<=50K           37.985714           42.338235  38.041667    36.030928   
>50K            42.440000           47.000000  48.750000    45.000000   

native-country    England     ...       Portugal  Puerto-Rico   Scotland  \
salary                        ...                                          
<=50K           40.483333     ...      41.939394    38.470588  39.444444   
>50K            44.533333     ...      41.500000    39.416667  46.666667   

native-country     South     Taiwan   Thailand  Trinadad&Tobago  \
salary                                                            
<=50K           40.15625  33.774194  42.866667        37.058824   
>50K            51.43750  46.800000  58.333333        40.000000   

native-country  United-States    Vietnam  Yugoslavia  
salary                                                
<=50K               38.799127  37.193548        41.6  
>50K                45.505369  39.200000        49.5  

[2 rows x 42 columns]